src/mcelog.c

```diff
@@ -0,0 +1,518 @@
+} mcelog_memory_rec_t;
+
+static int socket_close(socket_adapter_t *self);
+static int socket_write(socket_adapter_t *self, const char *msg,
+                        const size_t len);
+static int socket_reinit(socket_adapter_t *self);
+static int socket_receive(socket_adapter_t *self, FILE **p_file);
+
+static mcelog_config_t g_mcelog_config = {
+    .logfile = "/var/log/mcelog", .tid = 0,
+};
+
+static socket_adapter_t socket_adapter = {
+    .sock_fd = -1,
+    .unix_sock =
+        {
+            .sun_family = AF_UNIX, .sun_path = "/var/run/mcelog-client",
+        },
+    .lock = PTHREAD_RWLOCK_INITIALIZER,
+    .close = socket_close,
+    .write = socket_write,
+    .reinit = socket_reinit,
+    .receive = socket_receive,
+};
+
+static _Bool mcelog_thread_running = 0;
+
+static int mcelog_config(oconfig_item_t *ci) {
+  for (int i = 0; i < ci->children_num; i++) {
+    oconfig_item_t *child = ci->children + i;
+    if (strcasecmp("McelogClientSocket", child->key) == 0) {
+      if (cf_util_get_string_buffer(child, socket_adapter.unix_sock.sun_path,
+                                    sizeof(socket_adapter.unix_sock.sun_path)) <
+          0) {
+        ERROR("%s: Invalid configuration option: \"%s\".", MCELOG_PLUGIN,
+              child->key);
+        return -1;
+      }
+    } else if (strcasecmp("McelogLogfile", child->key) == 0) {
+      if (cf_util_get_string_buffer(child, g_mcelog_config.logfile,
+                                    sizeof(g_mcelog_config.logfile)) < 0) {
+        ERROR("%s: Invalid configuration option: \"%s\".", MCELOG_PLUGIN,
+              child->key);
+        return -1;
+      }
+    } else {
+      ERROR("%s: Invalid configuration option: \"%s\".", MCELOG_PLUGIN,
+            child->key);
+      return -1;
+    }
+  }
+  return (0);
+}
+
+static int socket_close(socket_adapter_t *self) {
+  int ret = 0;
+  pthread_rwlock_rdlock(&self->lock);
+  if (fcntl(self->sock_fd, F_GETFL) != -1) {
+    if (shutdown(self->sock_fd, SHUT_RDWR) != 0) {
+      char errbuf[MCELOG_BUFF_SIZE];
+      ERROR("%s: Socket shutdown failed: %s", MCELOG_PLUGIN,
+            sstrerror(errno, errbuf, sizeof(errbuf)));
+      ret = -1;
+    }
+    close(self->sock_fd);
+  }
+  pthread_rwlock_unlock(&self->lock);
+  return ret;
+}
+
+static int socket_write(socket_adapter_t *self, const char *msg,
+                        const size_t len) {
+  int ret = 0;
+  pthread_rwlock_rdlock(&self->lock);
+  if (swrite(self->sock_fd, msg, len) < 0)
+    ret = -1;
+  pthread_rwlock_unlock(&self->lock);
+  return ret;
+}
+
+static int socket_reinit(socket_adapter_t *self) {
+  char errbuff[MCELOG_BUFF_SIZE];
+  int flags;
+  int ret = -1;
+  cdtime_t interval = plugin_get_interval();
+  struct timeval socket_timeout = CDTIME_T_TO_TIMEVAL(interval);
+
+  /* synchronization via write lock since sock_fd may be changed here */
+  pthread_rwlock_wrlock(&self->lock);
+  self->sock_fd = socket(PF_UNIX, SOCK_STREAM, 0);
```

Comment:
regarding SOCK_NONBLOCK please see [d212c7f](https://github.com/collectd/collectd/pull/2003/commits/d212c7f2fb8af6d3ade42aaf1d019c5a2d7e8d3b). It failed on epel5